substantial/types.py

```diff
@@ -111,17 +144,44 @@ async def exec(self, ctx, save_id, counter) -> Any:
                 ret = op
             else:
                 raise Exception(
-                    f"Expected value or coroutine object, got {type(op)} instead"
+                    f"Expected value or coroutielf.ne object, got {type(op)} instead"
                 )
 
             save = events.Save(save_id, json.dumps(ret), -1)
             ctx.source(events.Event(save=save))
             return ret
         except Exception as e:
+            if compensation_stack and len(compensation_stack):
+                compensation_stack.reverse()
+                for compensation_fn in compensation_stack:
+                    try:
+                        compensation_result = compensation_fn()
+                        if inspect.iscoroutine(compensation_result):
+                            compensation_result = await compensation_result
+                        ctx.source(
+                            events.Event(
+                                compensation=events.Compensation(
+                                    save_id=save_id,
+                                    error=str(e),
+                                    compensation_result=json.dumps(
+                                        compensation_result,
+                                    ),
+                                )
+                            )
+                        )
+                    except Exception as compensation_error:
+                        raise CompensationFailed(e, compensation_error)
+
+                raise CancelWorkflow(f"Compensation completed for error: {e}")
```

Comment:
Maybe just rise the actual exception?

Reply:
Or just make sure we throw the same exception as a save that has failed without a compensate fn.